### PR TITLE
Remove python2.7 support 

### LINKED ISSUE #1395: Drop Python2.7 support
As noted by #1094, VFX platform 2019 is the last VFX platform that includes python 2.7.  This issue is to prepare for when we drop python 2.7 support.

- [ ] update check in setup.py: 
  - [ ] https://github.com/AcademySoftwareFoundation/OpenTimelineIO/blob/78941ddf12a16e6b3495290e7c920ce674b1f342/setup.py#L186
  - [ ] https://github.com/AcademySoftwareFoundation/OpenTimelineIO/blob/78941ddf12a16e6b3495290e7c920ce674b1f342/setup.py#L295
  - [ ] https://github.com/AcademySoftwareFoundation/OpenTimelineIO/blob/78941ddf12a16e6b3495290e7c920ce674b1f342/setup.py#L336
- [ ] update README.md
- [ ] update files that reference 2.7, especially places marked `# python2` or `# python 2.7`  (this is a rough list from grep)
```
tests/test_builtin_adapters.py
tests/test_console.py
tests/test_timeline.py
tests/test_cmx_3600_adapter.py
tests/test_otioz.py
tests/test_adapter_plugin.py
tests/baselines/plugin_module/otio_mockplugin/__init__.py
tests/test_hooks_plugins.py
tests/test_examples.py
contrib/opentimelineio_contrib/adapters/tests/test_hls_playlist_adapter.py
src/py-opentimelineio/opentimelineio/adapters/file_bundle_utils.py
src/py-opentimelineio/opentimelineio/adapters/otiod.py
src/py-opentimelineio/opentimelineio/adapters/fcp_xml.py
src/py-opentimelineio/opentimelineio/url_utils.py
```
- [ ] update `.github/workflows/python-package.yml` - [python-package.yml](https://github.com/AcademySoftwareFoundation/OpenTimelineIO/blob/main/.github/workflows/python-package.yml)
- [ ] update pybind11 to 2.10+
- [ ] try [pyupgrade](https://github.com/asottile/pyupgrade) to move to newer syntax where available

### PR DESCRIPTION
**Pull request by ssteinbach**
## Summary

Following the VFX platform, for which 2019 was the last version which included python 2.7 support, we are dropping python2.7 support starting with v0.16.  This enables us to modernize the code base and remove a number of shims and bandaids.

## Changes

* Sets VFX platform supported years as 2022, 2021, and 2020
* Removes Python2.7 support from the library
  * `# python 2` blocks
  * `version_info[0] < 3` blocks
  * `try:...except ImportError:... ` blocks
* Removes Python2.7 build matrix lines from CI/github actions
* updates setup.py project metadata to drop 2.7 from versions
* Removes shim code that was bridging differences between python3 and python2
* modernizes to python3 patterns by running `pyupgrade --py37-plus`
  * makes a bunch of smaller changes, from visual inspection lots of `"".format()` -> `f""`

## OTIOZ Note
We will follow this up with a PR proposing to change the ZIP_DEFLATED algorithm used by OTIOZ to use bzip2 or LZMA.

Closes #1395

### CODE REVIEW

**Comment by ssteinbach** [comment]
Review notes on a previous PR that I want to not lose track of: https://github.com/AcademySoftwareFoundation/OpenTimelineIO/pull/1418#issuecomment-1258478479

File: setup.py [hunk 1]
```diff
@@ -341,15 +339,10 @@ def test_otio():
     },
 
     # Disallow 3.9.0 because of https://github.com/python/cpython/pull/22670
-    python_requires='>2.7, !=3.0.*, !=3.1.*, !=3.2.*, !=3.3.*, !=3.4.*, !=3.5.*, !=3.6.*, !=3.9.0',  # noqa: E501
+    python_requires='>3.6, !=3.9.0',  # noqa: E501
```
Comment (JeanChristopheMorinPerso) [inline setup.py:342]:
```suggestion
    python_requires='>=3.7, !=3.9.0',  # noqa: E501
```

`>3.6` means that 3.6.1 would be accepted:
```python
>>> import packaging.specifiers
>>> spec = packaging.specifiers.SpecifierSet('>3.6')
>>> spec.contains('3.6.1')
True
```

**Comment by ssteinbach** [comment]
@JeanChristopheMorinPerso It looks like `importlib.metadata` is new in 3.8 - since we still have 3.7 on the list, I don't think we can use that quite yet.

**Comment by JeanChristopheMorinPerso** [comment]
@ssteinbach there is a backport on PyPI: https://pypi.org/project/importlib-metadata/. It's what I would call an official backport.

**Comment by ssteinbach** [comment]
@JeanChristopheMorinPerso Ok got it.  I think I'd like to split that out to a follow up PR.

**Review by JeanChristopheMorinPerso — COMMENTED** [review]
Looks good to me. I just found one small thing to fix. Also, should we replace the `next` methods for `__next__`? For example https://github.com/AcademySoftwareFoundation/OpenTimelineIO/blob/main/src/py-opentimelineio/opentimelineio-bindings/otio_serializableObjects.cpp#L247.

There is also some leftover python 2 vs 3 in `_core_utils.py`:
* https://github.com/AcademySoftwareFoundation/OpenTimelineIO/blob/main/src/py-opentimelineio/opentimelineio/core/_core_utils.py#L39-L69
* https://github.com/AcademySoftwareFoundation/OpenTimelineIO/blob/main/src/py-opentimelineio/opentimelineio/core/_core_utils.py#L234
* https://github.com/AcademySoftwareFoundation/OpenTimelineIO/blob/main/src/py-opentimelineio/opentimelineio/core/_core_utils.py#L385

File: .github/workflows/python-package.yml [hunk 2]
```diff
@@ -172,22 +169,6 @@ jobs:
...
-        if: matrix.python-build == 'cp27*'
-        # cibuildwheel 1.12.0 is the last release that supported Python 2.7.
-        uses: pypa/cibuildwheel@v1.12.0
-        with:
-          output-dir: wheelhouse
-        env:
-          CIBW_BUILD: ${{ matrix.python-build }}
-
       - name: Build wheels (Python 3)
         uses: pypa/cibuildwheel@v2.3.1
```
[the remark this answers is not in the document]
Comment (ssteinbach) [reply .github/workflows/python-package.yml:173]: yes: #1420

...but that requires getting the dependabot branch permissions untangled.

**Comment by ssteinbach** [comment]
> Looks good to me. I just found one small thing to fix. Also, should we replace the `next` methods for `__next__`? For example https://github.com/AcademySoftwareFoundation/OpenTimelineIO/blob/main/src/py-opentimelineio/opentimelineio-bindings/otio_serializableObjects.cpp#L247.
> 
> There is also some leftover python 2 vs 3 in `_core_utils.py`:
> 
> * https://github.com/AcademySoftwareFoundation/OpenTimelineIO/blob/main/src/py-opentimelineio/opentimelineio/core/_core_utils.py#L39-L69
> * https://github.com/AcademySoftwareFoundation/OpenTimelineIO/blob/main/src/py-opentimelineio/opentimelineio/core/_core_utils.py#L234
> * https://github.com/AcademySoftwareFoundation/OpenTimelineIO/blob/main/src/py-opentimelineio/opentimelineio/core/_core_utils.py#L385

Great catch!  This led me to a bunch of other ones by searching for `version_info` shenanigans.  Fixing shortly...

File: contrib/opentimelineio_contrib/adapters/advanced_authoring_format.py [hunk 3]
```diff
@@ -101,7 +90,7 @@ def _get_class_name(item):
 def _transcribe_property(prop, owner=None):
     # XXX: The unicode type doesn't exist in Python 3 (all strings are unicode)
     # so we have to use type(u"") which works in both Python 2 and 3.
-    if isinstance(prop, (str, type(u""), numbers.Integral, float, dict)):
+    if isinstance(prop, (str, str, numbers.Integral, float, dict)):
```
Comment (JeanChristopheMorinPerso) [inline contrib/opentimelineio_contrib/adapters/advanced_authoring_format.py:93]:
```suggestion
    if isinstance(prop, (str, numbers.Integral, float, dict)):
```

**Comment by ssteinbach** [comment]
Hm, I'm not sure why `collections.abc` classes are throwing Sphinx off, or why they're creeping into the doc build?

File: src/py-opentimelineio/opentimelineio/core/_core_utils.py [hunk 4]
```diff
@@ -372,22 +330,22 @@ def insert(self, index, item):
...
-                    setattr(sequenceClass, name, _im_func(func))
-                    if name.startswith('__') or name.endswith('__') or sys.version_info[0] < 3:  # noqa
+                    setattr(sequenceClass, name, func)
+                    if name.startswith('__') or name.endswith('__'):
                         continue
 
                     # Hide the method frm Sphinx doc.
                     # See https://www.sphinx-doc.org/en/master/usage/restructuredtext/domains.html#info-field-lists  # noqa
-                    getattr(sequenceClass, name).__doc__ += '\n\n:meta private:'
+                    getattr(sequenceClass, name).__doc__ += '\()n\n:meta private:'
```
Comment (JeanChristopheMorinPerso) [inline src/py-opentimelineio/opentimelineio/core/_core_utils.py]: I think this might be an unintentional change?

**Comment by JeanChristopheMorinPerso** [comment]
@ssteinbach I guess you found the issue (the extra `()`)?

**Review by JeanChristopheMorinPerso — APPROVED** [review]
I re-reviewed everything and it looks good to me!

I'll just ask a question again because I think it kind of got lost. Should we replace the `next` methods for `__next__`? For example https://github.com/AcademySoftwareFoundation/OpenTimelineIO/blob/main/src/py-opentimelineio/opentimelineio-bindings/otio_serializableObjects.cpp#L247. I think it would be a good idea but can well be push to another PR since this PR is starting to be pretty big.

**Review by meshula — APPROVED** [review]

File: .github/workflows/python-package.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (meshula) [reply .github/workflows/python-package.yml:173]: marking as resolved because the dependabot question does not block this merge

**Comment by ssteinbach** [comment]
Ok this looks good.  The two follow up from this:

- #1426 
- @JeanChristopheMorinPerso comments about `next`->`__next__`

**Pull request merged**